### PR TITLE
[python] Process ZCL responses

### PR DESCRIPTION
**Pull request by Damian-Nordic**
#### Problem
Python CHIP controller `zcl` and `zclread` commands send ZCL requests to a peer device, but responses to the requests are not processed nor displayed to a user and using especially the `zclread` command seems pointless.

 #### Summary of Changes
1. Move DataModelHandler from `app/server` to `app/util`. DataModelHandler is used on both controller and accessory side, so it makes more sense to keep it in the util directory.
2. Make Python CHIP controller process ZCL responses. Generate ZCL code for Python CHIP Controller and build    DataModelHandler in the context of the generated code.
3. Implement ZCL response callbacks. Implement ZCL response callbacks for commands and attribute read requests so that contents of a response is printed to the console. Make Python code provide success and failure delegates which complete an asynchronous operation and return to the prompt.
4. Provide empty DataModelHandler functions.
5. Regenerate Python CHIP Controller gen/ in the common script

Usage example (getting state of the OnOff cluster):
```
chip-device-ctrl > zclread OnOff OnOff 1 1 1
CHIP:ZCL: Successfully encoded 11 bytes
CHIP:IN: Secure message was encrypted: Msg ID 4
CHIP:IN: Sending msg from 0x000000000001b669to 0x0000000000000001
CHIP:IN: Sending secure msg on generic transport
CHIP:IN: Secure msg send status 0
CHIP:CTL: SendMessage returned 0
CHIP:ZCL: APS frame processing success!
CHIP:ZCL: RX len 8, ep 1, clus 0x 6 
CHIP:ZCL:  FC 8 seq 4 cmd 1 payload[
CHIP:ZCL: 00 00 00 10 01 
CHIP:ZCL: ]
CHIP:ZCL: ReadAttributesResponse:
CHIP:ZCL:   ClusterId: 0x0006
CHIP:ZCL:   attributeId: 0x0000
CHIP:ZCL:   status: EMBER_ZCL_STATUS_SUCCESS (0x00)
CHIP:ZCL:   attributeType: 0x10
CHIP:ZCL:   value: 1
Secure Session to Device Established
CHIP:ZCL: Data model processing success!
chip-device-ctrl >
```
**Attention**
Files in `gen/` directory, `clusters/CHIPClusters.cpp` and `clusters/CHIPClusters.py` are auto-generated and don't need to be reviewed.

### CODE REVIEW

**Comment by cjandhyala** [comment]
Can you add time stamp to the msgs sent/received on the python controller side ?

**Comment by Damian-Nordic** [comment]
> Can you add time stamp to the msgs sent/received on the python controller side ?

The messages are printed by the common CHIP code, but I've seen some changes made by Andy to redirect CHIP log messages to Python logging subsystem so that any format can be applied. Is that a requirement for TE2? I can log an issue for that.

**Comment by andy31415** [comment]
Python code has the ability to redirect logging and the python scripts can control outputs (e.g. add timestamps to them)

File: examples/all-clusters-app/all-clusters-common/BUILD.gn [hunk 1]
```diff
@@ -53,7 +53,7 @@ source_set("all-clusters-common") {
     "${chip_root}/src/app/clusters/tv-channel-server/tv-channel-server.cpp",
     "${chip_root}/src/app/reporting/reporting-default-configuration.cpp",
     "${chip_root}/src/app/reporting/reporting.cpp",
-    "${chip_root}/src/app/server/DataModelHandler.cpp",
+    "${chip_root}/src/app/util/DataModelHandler.cpp",
```
Comment (andy31415) [inline examples/all-clusters-app/all-clusters-common/BUILD.gn:56]: why the move from server to util?

generally I am reluctant to place things under util/helper/support named things as it becomes a catch-all for everything.
Are we removing 'server' completely? what would be the intended final layout of files?

**Review by andy31415 — APPROVED** [review]
I am unable to code review 40K lines PRs.

Rubber stamping as code tht we need (want ZCL replies for python) but very worried about our PR sizes.

**Comment by andy31415** [comment]
@Damian-Nordic - could you highlight which files are not robot-generated and could be reviewed?

**Comment by Damian-Nordic** [comment]
> @Damian-Nordic - could you highlight which files are not robot-generated and could be reviewed?

In general (applies to all PRs), all files in `gen/` directories are auto-generated from ZAP templates. In my PR, `clusters/CHIPClusters.cpp` and `clusters/CHIPClusters.py` are auto-generated, too. You can see the list of auto-generated files in `src/app/zap-templates/chip-templates.json` (CHIP common files) and `src/app/zap-templates/app-templates.json` (application specific files).

File: examples/all-clusters-app/all-clusters-common/BUILD.gn (hunk 1 above)
Comment (Damian-Nordic) [reply examples/all-clusters-app/all-clusters-common/BUILD.gn:56]: Having it in `src/app/server` doesn't make much sense for several reasons:
1) all other files in `src/app/server` are built as a library defined in `src/app/server/BUILD.gn` while `DataModelHandler.cpp` is included directly by applications (like other Data Model files from `src/app/util`)
2) other files in `src/app/server` are only relevant for the accessory/server side while `DataModelHandler.cpp` is pulled by both controllers and accessories.
If found it quite ugly that we pulled files from `src/app/server` into controller applications to handle client side of the ZCL clusters.

**Review by msandstedt — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Pull request merged**